### PR TITLE
Add CTC loss option in zipformer recipe

### PR DESCRIPTION
**Pull request by yaozengwei**
This PR adds CTC loss option in the new zipformer recipe. Now `train.py` supports: 

- transducer loss (default), with `--use-transducer=True --use-ctc=False`
- ctc loss, with `--use-transducer=False --use-ctc=True`
- transducer loss & ctc loss, with `--use-transducer=True --use-ctc=True`

The change dose not cause conflicts with already trained model.

### CODE REVIEW

**Comment by yaozengwei** [comment]
Training with both transducer loss and CTC loss, with option `--use-transducer True --use-ctc True`

* transducer head, epoch-40-avg-16

| greedy-search | modified-beam-search | fast-beam-search | 
| - | - | - | 
| 2.24, 5.01 | 2.19, 5.03 | 2.2, 4.98 |

* CTC head, epoch-40-avg-16

| ctc-decoding | 1best | nbest | nbest-rescoring | whole-lattice-rescoring | 
| - | - | - | - | - |
| 2.4, 5.66 | 2.46, 5.11 | 2.46, 5.11 | 2.37, 4.93 | 2.37, 4.88 |

@marcoyang1998 has trained the zipformer with CTC loss alone, getting 2.62, 6.14 at epoch-90-avg-38.

**Comment by yaozengwei** [comment]
> What is the graph and lm?
For 1best and nbest, we use HLG as decoding graph, where G is a word-level 3-gram. For nbest-rescoring and whole-lattice-rescoring, we use a word-level 4-gram for further rescoring.

File: egs/librispeech/ASR/zipformer/model.py [hunk 1]
```diff
@@ -214,4 +273,86 @@ def forward(
...
+                am_scale=am_scale,
+                lm_scale=lm_scale,
+            )
+        else:
+            simple_loss = 0
+            pruned_loss = 0
+
+        if self.use_ctc:
+            # Compute CTC loss
+            targets = [t for tokens in y.tolist() for t in tokens]
```
Comment (csukuangfj) [inline egs/librispeech/ASR/zipformer/model.py:345]:
```suggestion
            targets = y.values
```

Please see https://k2-fsa.github.io/k2/python_api/api.html#values

**Comment by csukuangfj** [comment]
> Training with both transducer loss and CTC loss, with option `--use-transducer True --use-ctc True`
> 
> * transducer head, epoch-40-avg-16
> 
> greedy-search	modified-beam-search	fast-beam-search
> 2.24, 5.01	2.19, 5.03	2.2, 4.98
> * CTC head, epoch-40-avg-16
> 
> ctc-decoding	1best	nbest	nbest-rescoring	whole-lattice-rescoring
> 2.4, 5.66	2.46, 5.11	2.46, 5.11	2.37, 4.93	2.37, 4.88
> @marcoyang1998 has trained the zipformer with CTC loss alone, getting 2.62, 6.14 at epoch-90-avg-38.

Could you list the baseline results trained without CTC for better comparison?

File: egs/librispeech/ASR/zipformer/pretrained_ctc.py [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+# limitations under the License.
+"""
+This script loads a checkpoint and uses it to decode waves.
+You can generate the checkpoint with the following command:
+
+- For non-streaming model:
+
+./zipformer/export.py \
+  --exp-dir ./zipformer/exp \
+  --use-ctc 1 \
```
Comment (csukuangfj) [inline egs/librispeech/ASR/zipformer/pretrained_ctc.py:26]: Shall we also set `--use-transducer 0` to discard the transducer branch? 

It can reduce the model size.

File: egs/librispeech/ASR/zipformer/pretrained_ctc.py [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+            the transformer encoder output is used for decoding.
+            We call it HLG decoding.
+        (2) nbest-rescoring. Extract n paths from the decoding lattice,
+            rescore them with an LM, the path with
+            the highest score is the decoding result.
+            We call it HLG decoding + n-gram LM rescoring.
+        (3) whole-lattice-rescoring - Use an LM to rescore the
+            decoding lattice and then use 1best to decode the
+            rescored lattice.
+            We call it HLG decoding + n-gram LM rescoring.
```
Comment (csukuangfj) [inline egs/librispeech/ASR/zipformer/pretrained_ctc.py:174]: It has the same name as `(2)`. Could you fix it?

File: egs/librispeech/ASR/zipformer/pretrained_ctc.py [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
+
+    parser.add_argument(
+        "--num-classes",
+        type=int,
+        default=500,
+        help="""
+        Vocab size in the BPE model.
+        """,
+    )
+
```
Comment (csukuangfj) [inline egs/librispeech/ASR/zipformer/pretrained_ctc.py:229]: This argument can be removed. We can read the vocab size from the given `bpe.model`.

File: egs/librispeech/ASR/zipformer/pretrained_ctc.py [hunk 5]
```diff
@@ -0,0 +1,453 @@
...
+      Return a list of 1-D float32 torch tensors.
+    """
+    ans = []
+    for f in filenames:
+        wave, sample_rate = torchaudio.load(f)
+        assert sample_rate == expected_sample_rate, (
+            f"expected sample rate: {expected_sample_rate}. " f"Given: {sample_rate}"
+        )
+        # We use only the first channel
+        ans.append(wave[0])
```
Comment (csukuangfj) [inline egs/librispeech/ASR/zipformer/pretrained_ctc.py:271]:
```suggestion
        ans.append(wave[0].contiguous())
```

File: egs/librispeech/ASR/zipformer/pretrained_ctc.py [hunk 6]
```diff
@@ -0,0 +1,453 @@
...
+
+    if params.method == "ctc-decoding":
+        logging.info("Use CTC decoding")
+        bpe_model = spm.SentencePieceProcessor()
+        bpe_model.load(params.bpe_model)
+        max_token_id = params.num_classes - 1
+
+        H = k2.ctc_topo(
+            max_token=max_token_id,
+            modified=False,
```
Comment (csukuangfj) [inline egs/librispeech/ASR/zipformer/pretrained_ctc.py:349]: I suggest passing `modified` as a commandline argument. For languages with a very large vocab size, we need to set it to `True` to avoid OOM.

File: egs/librispeech/ASR/zipformer/pretrained_ctc.py [hunk 7]
```diff
@@ -0,0 +1,453 @@
...
+            best_path_dict = rescore_with_n_best_list(
+                lattice=lattice,
+                G=G,
+                num_paths=params.num_paths,
+                lm_scale_list=[params.ngram_lm_scale],
+                nbest_scale=params.nbest_scale,
+            )
+            best_path = next(iter(best_path_dict.values()))
+        elif params.method == "whole-lattice-rescoring":
+            logging.info("Use HLG decoding + LM rescoring")
```
Comment (csukuangfj) [inline egs/librispeech/ASR/zipformer/pretrained_ctc.py:426]: It is the same as line `416`. Please fix it.

File: egs/librispeech/ASR/zipformer/jit_pretrained_ctc.py [hunk 8]
```diff
@@ -0,0 +1,434 @@
...
+# limitations under the License.
+"""
+This script loads a checkpoint and uses it to decode waves.
+You can generate the checkpoint with the following command:
+
+- For non-streaming model:
+
+./zipformer/export.py \
+  --exp-dir ./zipformer/exp \
+  --use-ctc 1 \
```
Comment (csukuangfj) [inline egs/librispeech/ASR/zipformer/jit_pretrained_ctc.py:26]:
```suggestion
  --use-ctc 1 \
  --use-transducer 0 \
```

**Comment by yaozengwei** [comment]
> > Training with both transducer loss and CTC loss, with option `--use-transducer True --use-ctc True`
> > 
> > * transducer head, epoch-40-avg-16
> > 
> > greedy-search	modified-beam-search	fast-beam-search
> > 2.24, 5.01	2.19, 5.03	2.2, 4.98
> > 
> > * CTC head, epoch-40-avg-16
> > 
> > ctc-decoding	1best	nbest	nbest-rescoring	whole-lattice-rescoring
> > 2.4, 5.66	2.46, 5.11	2.46, 5.11	2.37, 4.93	2.37, 4.88
> > @marcoyang1998 has trained the zipformer with CTC loss alone, getting 2.62, 6.14 at epoch-90-avg-38.
> 
> Could you list the baseline results trained without CTC for better comparison?

Added

File: egs/librispeech/ASR/zipformer/model.py [hunk 9]
```diff
@@ -55,47 +60,130 @@ def __init__(
...
+          use_transducer:
+            Whether use transducer head. Default: True.
+          use_ctc:
+            Whether use CTC head. Default: False.
         """
         super().__init__()
+
+        assert (
+            use_transducer or use_ctc
+        ), f"At least one of them should be True, but gotten use_transducer={use_transducer}, use_ctc={use_ctc}"
```
[the remark this answers is not in the document]
Comment (yaozengwei) [reply egs/librispeech/ASR/zipformer/model.py:78]: Thanks
